[SQL] RustError: `select ascii(chr(<nullable int column>))` due to expected `i32`, but found `Option<i32>`

SQL:

```sql
CREATE TABLE t0(c0 INT);
CREATE MATERIALIZED VIEW v20_optimized AS (SELECT ASCII(CHR(t0.c0)) FROM t0);
```

Fails with the following Program Error:

```rust
Rust error: the Rust code generated based on the SQL failed to compile, which should not happen. Please file a bug report with the example SQL that triggers it at: https://github.com/feldera/feldera/issues

The compilation task produced the following output:

stdout:

stderr:
   Compiling project01938c4e-0e57-7650-ad6c-e0332ae987d9 v0.1.0 (/home/abhizer/.feldera/cargo_workspace/project01938c4e-0e57-7650-ad6c-e0332ae987d9)
error[E0308]: mismatched types
   --> project01938c4e-0e57-7650-ad6c-e0332ae987d9/src/main.rs:127:23
    |
127 |             Tup1::new(asciiN(chrN((*t_1).0)))
    |             --------- ^^^^^^^^^^^^^^^^^^^^^^ expected `i32`, found `Option<i32>`
    |             |
    |             arguments to this function are incorrect
    |
    = note: expected type `i32`
               found enum `std::option::Option<i32>`
note: associated function defined here
   --> /mnt/hdd0/Development/feldera.git/main/crates/dbsp/src/utils/tuple/mod.rs:16:1
    |
16  | / crate::declare_tuples! {
17  | |     Tup1<T1>,
18  | |     Tup2<T1, T2>,
19  | |     Tup3<T1, T2, T3>,
...   |
26  | |     Tup10<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10>,
27  | | }
    | |_^
    = note: this error originates in the macro `crate::declare_tuples` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider using `Option::expect` to unwrap the `std::option::Option<i32>` value, panicking if the value is an `Option::None`
    |
127 |             Tup1::new(asciiN(chrN((*t_1).0)).expect("REASON"))
    |                                             +++++++++++++++++

For more information about this error, try `rustc --explain E0308`.
error: could not compile `project01938c4e-0e57-7650-ad6c-e0332ae987d9` (bin "project01938c4e-0e57-7650-ad6c-e0332ae987d9") due to 1 previous error


Exit code: exit status: 101
```

--- a/crates/sqllib/src/casts.rs
+++ b/crates/sqllib/src/casts.rs
@@ -1533,7 +1533,7 @@ macro_rules! cast_ts {
 
             cast_function!($type_name, $arg_type, Timestamp, Timestamp);
         }
-    }
+    };
 }
 
 cast_ts!(i32, i32);

--- a/crates/sqllib/src/casts.rs
+++ b/crates/sqllib/src/casts.rs
@@ -1498,6 +1498,50 @@ pub fn cast_to_Timestamp_Timestamp(value: Timestamp) -> Timestamp {
 
 cast_function!(Timestamp, Timestamp, Timestamp, Timestamp);
 
+#[doc(hidden)]
+#[inline]
+pub fn cast_to_Timestamp_i64(value: i64) -> Timestamp {
+    Timestamp::new(value)
+}
+
+cast_function!(Timestamp, Timestamp, i64, i64);
+
+#[doc(hidden)]
+#[inline]
+pub fn cast_to_i64_Timestamp(value: Timestamp) -> i64 {
+    value.milliseconds()
+}
+
+cast_function!(i64, i64, Timestamp, Timestamp);
+
+macro_rules! cast_ts {
+    ($type_name: ident, $arg_type: ty) => {
+        ::paste::paste! {
+            #[doc(hidden)]
+            #[inline]
+            pub fn [<cast_to_Timestamp_ $type_name>](value: $arg_type) -> Timestamp {
+                cast_to_Timestamp_i64([< cast_to_i64_ $type_name >](value))
+            }
+
+            cast_function!(Timestamp, Timestamp, $type_name, $arg_type);
+
+            #[doc(hidden)]
+            #[inline]
+            pub fn [<cast_to_ $type_name _Timestamp>](value: Timestamp) -> $arg_type {
+                [< cast_to_ $type_name _i64 >] (cast_to_i64_Timestamp(value))
+            }
+
+            cast_function!($type_name, $arg_type, Timestamp, Timestamp);
+        }
+    }
+}
+
+cast_ts!(i32, i32);
+cast_ts!(i16, i16);
+cast_ts!(i8, i8);
+cast_ts!(f, F32);
+cast_ts!(d, F64);
+
 //////////////////// Other casts
 
 #[doc(hidden)]

--- a/crates/sqllib/src/timestamp.rs
+++ b/crates/sqllib/src/timestamp.rs
@@ -249,6 +249,14 @@ impl Timestamp {
         // Is this right for negative timestamps?
         Date::new((self.milliseconds / 86_400_000i64) as i32)
     }
+
+    /// Check if the timestamp is legal; panic if it isn't
+    pub fn check_legal(&self) {
+        let date = self.to_naiveDateTime();
+        if date.year() < 1 || date.year() > 9999 {
+            panic!("Timestamp out of range");
+        }
+    }
 }
 
 #[doc(hidden)]

--- a/docs/sql/datetime.md
+++ b/docs/sql/datetime.md
@@ -160,6 +160,11 @@ The following operations are available on timestamps:
 
 ### Operations on timestamps
 
+A cast from a numeric value to a `TIMESTAMP` interprets the numeric
+value as an (big integer) number of milliseconds since the Unix epoch.
+Conversely, a cast from a `TIMESTAMP` to a numeric value retrieves the
+number of milliseconds since the Unix epoch from the timestamp.
+
 `DATE_TRUNC(timestamp, <unit>)`, where `<unit>` is a time unit, as
 described above, between `MILLENNIUM` and `DAY`.  Result is a
 `TIMESTAMP`.  Rounds down the timestamp to the specified time unit.

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/DBSPCompiler.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/DBSPCompiler.java
@@ -561,36 +561,41 @@ public class DBSPCompiler implements IWritesLogs, ICompilerComponent, IErrorRepo
             } else {
                 this.messages.reportError(e);
             }
-            if (this.options.languageOptions.throwOnError) {
-                this.printMessages(currentView);
-                throw new RuntimeException(e);
-            }
+            this.rethrow(new RuntimeException(e), currentView);
         } catch (CalciteContextException e) {
             this.messages.reportError(e);
-            if (this.options.languageOptions.throwOnError) {
-                this.printMessages(currentView);
-                throw new RuntimeException(e);
-            }
+            this.rethrow(e, currentView);
         } catch (CalciteException e) {
             this.messages.reportError(e);
-            if (this.options.languageOptions.throwOnError) {
-                this.printMessages(currentView);
-                throw new RuntimeException(e);
-            }
+            this.rethrow(e, currentView);
         } catch (BaseCompilerException e) {
             this.messages.reportError(e);
-            if (this.options.languageOptions.throwOnError) {
-                this.printMessages(currentView);
-                throw e;
+            this.rethrow(e, currentView);
+        } catch (RuntimeException e) {
+            Throwable current = e;
+            while (current.getCause() != null) {
+                // Try to find a nicer cause
+                Throwable t = current.getCause();
+                if (t instanceof CalciteException ex) {
+                    this.messages.reportError(ex);
+                    this.rethrow(ex, currentView);
+                }
+                current = t;
             }
+            this.messages.reportError(e);
+            this.rethrow(e, currentView);
         } catch (Throwable e) {
             this.messages.reportError(e);
-            if (this.options.languageOptions.throwOnError) {
-                this.printMessages(currentView);
-                throw e;
-            }
+            this.rethrow(new RuntimeException(e), currentView);
         } finally {
             this.toCompile.clear();
+        }
+    }
+
+    void rethrow(RuntimeException e, @Nullable CreateViewStatement currentView) {
+        if (this.options.languageOptions.throwOnError) {
+            this.printMessages(currentView);
+            throw e;
         }
     }
 

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustVisitor.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustVisitor.java
@@ -75,6 +75,7 @@ import org.dbsp.sqlCompiler.compiler.frontend.statements.IHasSchema;
 import org.dbsp.sqlCompiler.compiler.visitors.VisitDecision;
 import org.dbsp.sqlCompiler.compiler.visitors.inner.EliminateStructs;
 import org.dbsp.sqlCompiler.compiler.visitors.inner.InnerVisitor;
+import org.dbsp.sqlCompiler.compiler.visitors.outer.CircuitRewriter;
 import org.dbsp.sqlCompiler.compiler.visitors.outer.CircuitVisitor;
 import org.dbsp.sqlCompiler.ir.IDBSPInnerNode;
 import org.dbsp.sqlCompiler.ir.IDBSPNode;
@@ -451,6 +452,17 @@ public class ToRustVisitor extends CircuitVisitor {
         for (IDBSPNode node : circuit.getAllOperators())
             if (node.is(DBSPSourceBaseOperator.class))
                 this.processNode(node);
+
+        FindComparators compFinder = new FindComparators(this.compiler);
+        FindStatics staticsFinder = new FindStatics(this.compiler);
+        compFinder.getCircuitVisitor(false).apply(circuit);
+        staticsFinder.getCircuitVisitor(false).apply(circuit);
+
+        for (DBSPComparatorExpression comparator: compFinder.found)
+            this.generateCmpFunc(comparator);
+        for (DBSPStaticExpression comparator: staticsFinder.found)
+            this.generateStatic(comparator);
+
         for (IDBSPNode node : circuit.getAllOperators())
             if (!node.is(DBSPSourceBaseOperator.class))
                 this.processNode(node);
@@ -891,18 +903,6 @@ public class ToRustVisitor extends CircuitVisitor {
 
     @Override
     public VisitDecision preorder(DBSPSimpleOperator operator) {
-        FindComparators compFinder = new FindComparators(this.compiler);
-        FindStatics staticsFinder = new FindStatics(this.compiler);
-        if (operator.function != null) {
-            compFinder.apply(operator.getFunction());
-            staticsFinder.apply(operator.getFunction());
-        }
-
-        for (DBSPComparatorExpression comparator: compFinder.found)
-            this.generateCmpFunc(comparator);
-        for (DBSPStaticExpression comparator: staticsFinder.found)
-            this.generateStatic(comparator);
-
         DBSPType streamType = new DBSPTypeStream(operator.outputType);
         this.writeComments(operator)
                 .append("let ")

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustVisitor.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustVisitor.java
@@ -75,7 +75,6 @@ import org.dbsp.sqlCompiler.compiler.frontend.statements.IHasSchema;
 import org.dbsp.sqlCompiler.compiler.visitors.VisitDecision;
 import org.dbsp.sqlCompiler.compiler.visitors.inner.EliminateStructs;
 import org.dbsp.sqlCompiler.compiler.visitors.inner.InnerVisitor;
-import org.dbsp.sqlCompiler.compiler.visitors.outer.CircuitRewriter;
 import org.dbsp.sqlCompiler.compiler.visitors.outer.CircuitVisitor;
 import org.dbsp.sqlCompiler.ir.IDBSPInnerNode;
 import org.dbsp.sqlCompiler.ir.IDBSPNode;

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/errors/BaseCompilerException.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/errors/BaseCompilerException.java
@@ -5,9 +5,7 @@ import org.dbsp.sqlCompiler.compiler.frontend.calciteObject.CalciteObject;
 
 import javax.annotation.Nullable;
 
-/**
- * Base class for exceptions which are thrown by the compiler.
- */
+/** Base class for exceptions which are thrown by the compiler. */
 public abstract class BaseCompilerException
         extends RuntimeException
         implements IHasSourcePositionRange {

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/AggregateCompiler.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/AggregateCompiler.java
@@ -40,6 +40,7 @@ import org.apache.calcite.sql.fun.SqlSumEmptyIsZeroAggFunction;
 import org.apache.calcite.util.ImmutableBitSet;
 import org.dbsp.sqlCompiler.compiler.DBSPCompiler;
 import org.dbsp.sqlCompiler.compiler.ICompilerComponent;
+import org.dbsp.sqlCompiler.compiler.errors.CompilationError;
 import org.dbsp.sqlCompiler.compiler.errors.UnimplementedException;
 import org.dbsp.sqlCompiler.compiler.frontend.calciteObject.CalciteObject;
 import org.dbsp.sqlCompiler.ir.aggregate.AggregateBase;
@@ -66,6 +67,7 @@ import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeBinary;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeBool;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeDouble;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeInteger;
+import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeNull;
 import org.dbsp.sqlCompiler.ir.type.user.DBSPTypeUser;
 import org.dbsp.sqlCompiler.ir.type.user.DBSPTypeVec;
 import org.dbsp.util.ICastable;
@@ -130,6 +132,10 @@ public class AggregateCompiler implements ICompilerComponent {
         } else if (argList.size() == 1) {
             int fieldNumber = call.getArgList().get(0);
             this.aggArgument = this.v.deref().field(fieldNumber);
+            if (this.aggArgument.getType().is(DBSPTypeNull.class)) {
+                throw new CompilationError("Argument of aggregate has NULL type",
+                        CalciteObject.create(call.getParserPosition()));
+            }
         } else {
             List<DBSPExpression> fields = Linq.map(call.getArgList(),
                     a -> this.v.deref().field(a).applyCloneIfNeeded());

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
@@ -1140,14 +1140,7 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression>
                         }
                         return compileFunction(module_prefix + getCallName(call), node, type, ops, 3, 4);
                     }
-                    case "chr": {
-                        // workaround https://issues.apache.org/jira/browse/CALCITE-6707
-                        validateArgCount(node, opName, ops.size(), 1);
-                        if (ops.get(0).getType().mayBeNull) {
-                            type = type.withMayBeNull(true);
-                        }
-                        // fall through
-                    }
+                    case "chr":
                     case "ascii":
                     case "lower":
                     case "upper":

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/CastTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/CastTests.java
@@ -28,10 +28,12 @@ import org.dbsp.sqlCompiler.compiler.frontend.calciteObject.CalciteObject;
 import org.dbsp.sqlCompiler.compiler.sql.tools.BaseSQLTests;
 import org.dbsp.sqlCompiler.compiler.sql.tools.Change;
 import org.dbsp.sqlCompiler.compiler.sql.tools.InputOutputChange;
+import org.dbsp.sqlCompiler.compiler.sql.tools.SqlIoTest;
 import org.dbsp.sqlCompiler.ir.expression.DBSPTupleExpression;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPDecimalLiteral;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPDoubleLiteral;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPI32Literal;
+import org.dbsp.sqlCompiler.ir.expression.literal.DBSPI64Literal;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPLiteral;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPStringLiteral;
 import org.dbsp.sqlCompiler.ir.expression.literal.DBSPZSetLiteral;
@@ -40,22 +42,21 @@ import org.junit.Test;
 
 import java.math.BigDecimal;
 
-public class CastTests extends BaseSQLTests {
+public class CastTests extends SqlIoTest {
     final DBSPTypeDecimal tenTwo = new DBSPTypeDecimal(CalciteObject.EMPTY, 10, 2, true);
     final DBSPTypeDecimal tenFour = new DBSPTypeDecimal(CalciteObject.EMPTY, 10, 4, false);
 
-    public DBSPCompiler compileQuery(String query) {
-        DBSPCompiler compiler = this.testCompiler();
+    @Override
+    public void prepareInputs(DBSPCompiler compiler) {
         String ddl = "CREATE TABLE T (\n" +
                 "COL1 INT NOT NULL" +
                 ", COL2 DOUBLE NOT NULL" +
                 ", COL3 VARCHAR NOT NULL" +
                 ", COL4 DECIMAL(10,2)" +
                 ", COL5 DECIMAL(10,4) NOT NULL" +
-                ")";
-        compiler.compileStatement(ddl);
-        compiler.compileStatement(query);
-        return compiler;
+                ");" +
+                "INSERT INTO T VALUES(10, 12.0, 100100, NULL, 100103);";
+        compiler.compileStatements(ddl);
     }
 
     public Change createInput() {
@@ -68,9 +69,8 @@ public class CastTests extends BaseSQLTests {
     }
 
     public void testQuery(String query, DBSPZSetLiteral expectedOutput) {
-        query = "CREATE VIEW V AS " + query;
-        DBSPCompiler compiler = this.compileQuery(query);
-        CompilerCircuitStream ccs = new CompilerCircuitStream(compiler);
+        query = "CREATE VIEW V AS " + query + ";";
+        CompilerCircuitStream ccs = this.getCCS(query);
         InputOutputChange change = new InputOutputChange(this.createInput(), new Change(expectedOutput));
         ccs.addChange(change);
         this.addRustTestCase(ccs);
@@ -119,5 +119,75 @@ public class CastTests extends BaseSQLTests {
         this.runtimeFail("SELECT CAST(100103123 AS DECIMAL(10, 4))",
                 "cannot represent 100103123 as DECIMAL(10, 4)",
                 this.streamWithEmptyChanges());
+    }
+
+    @Test
+    public void testFpCasts() {
+        this.testQuery("SELECT CAST(T.COL2 AS BIGINT) FROM T", new DBSPZSetLiteral(new DBSPTupleExpression(new DBSPI64Literal(12))));
+    }
+
+    @Test
+    public void timeCastTests() {
+        this.qs("""
+                SELECT CAST(1000 AS TIMESTAMP);
+                 t
+                ---
+                 1970-01-01 00:00:01
+                (1 row)
+                
+                SELECT CAST(3600000 AS TIMESTAMP);
+                 t
+                ---
+                 1970-01-01 01:00:00
+                (1 row)
+                
+                SELECT CAST(-1000 AS TIMESTAMP);
+                 t
+                ---
+                 1969-12-31 23:59:59
+                (1 row)
+                
+                SELECT CAST(TIMESTAMP '1970-01-01 00:00:01.234' AS INTEGER);
+                 i
+                ---
+                 1234
+                (1 row)
+                
+                SELECT CAST(T.COL1 * 1000 AS TIMESTAMP) FROM T;
+                 t
+                ---
+                 1970-01-01 00:00:10
+                (1 row)
+                
+                SELECT CAST(T.COL2 * 1000 AS TIMESTAMP) FROM T;
+                 t
+                ---
+                 1970-01-01 00:00:12
+                (1 row)
+                
+                SELECT CAST(CAST(T.COL2 * 1000 AS TIMESTAMP) AS INTEGER) FROM T;
+                 i
+                ---
+                 12000
+                (1 row)
+                
+                SELECT CAST(CAST(T.COL2 / 10 AS TIMESTAMP) AS DOUBLE) FROM T;
+                 i
+                ---
+                 1
+                (1 row)""");
+    }
+
+    @Test
+    public void testFailingTimeCasts() {
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT CAST(1000 AS TIME)",
+                "Cast function cannot convert value of type INTEGER to type TIME");
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT CAST(TIME '10:00:00' AS INTEGER)",
+                "Cast function cannot convert value of type TIME(0) to type INTEGER");
+
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT CAST(1000 AS DATE)",
+                "Cast function cannot convert value of type INTEGER to type DATE");
+        this.statementsFailingInCompilation("CREATE VIEW V AS SELECT CAST(DATE '2024-01-01' AS INTEGER)",
+                "Cast function cannot convert value of type DATE to type INTEGER");
     }
 }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/OptimizedVariantTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/OptimizedVariantTests.java
@@ -8,8 +8,6 @@ public class OptimizedVariantTests extends VariantTests {
     @Override
     public DBSPCompiler testCompiler() {
         CompilerOptions options = this.testOptions(false, true);
-        DBSPCompiler compiler = new DBSPCompiler(options);
-        this.prepareInputs(compiler);
-        return compiler;
+        return new DBSPCompiler(options);
     }
 }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
@@ -52,8 +52,9 @@ public class RegressionTests extends SqlIoTest {
 
     @Test
     public void issue3076() {
-        this.compileRustTestCase("""
-                CREATE MATERIALIZED VIEW v72_optimized AS SELECT SUM(NULL);""");
+        this.statementsFailingInCompilation("CREATE VIEW v AS SELECT SUM(NULL);",
+                "Argument of aggregate has NULL type");
+        this.compileRustTestCase("CREATE VIEW V AS SELECT SUM(CAST(NULL AS INTEGER));");
     }
 
     @Test

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
@@ -98,7 +98,7 @@ public class RegressionTests extends SqlIoTest {
     @Test
     public void issue3041() {
         this.statementsFailingInCompilation("CREATE MATERIALIZED VIEW v199 AS SELECT TRIM();",
-                "Expression is missing some required operands");
+                "Error parsing SQL: Encountered \")\" at line 1");
     }
 
     @Test

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
@@ -15,6 +15,50 @@ import org.junit.Test;
 
 public class RegressionTests extends SqlIoTest {
     @Test
+    public void issue3070() {
+        this.compileRustTestCase("""
+                CREATE TABLE t0(c0 INT);
+                CREATE MATERIALIZED VIEW v20_optimized AS (SELECT ASCII(CHR(t0.c0)) FROM t0);""");
+    }
+
+    @Test
+    public void issue3071() {
+        this.compileRustTestCase("""
+                CREATE TABLE t0(c0 char) with ('materialized' = 'true');
+                CREATE MATERIALIZED VIEW v100_optimized AS (SELECT * FROM t0 WHERE (t0.c0) NOT BETWEEN ('Y') AND ('䤈'));""");
+    }
+
+    @Test
+    public void issue3072() {
+        this.compileRustTestCase("""
+                CREATE TABLE t0(c0 VARCHAR);
+                CREATE VIEW v271_optimized AS
+                SELECT COUNT(RLIKE(t0.c0, '0.5123590946084831')) FROM t0;""");
+        this.compileRustTestCase("""
+                CREATE TABLE t0(c0 DOUBLE, c1 INT, c2 BIGINT);
+                CREATE MATERIALIZED VIEW v293_optimized AS SELECT COUNT(RLIKE(SUBSTRING('', t0.c1), '')) FROM t0;""");
+    }
+
+    @Test
+    public void issue3073() {
+        this.compileRustTestCase("""
+                CREATE MATERIALIZED VIEW v10_optimized AS
+                SELECT SOME(('2059-10-5 5:3:37') NOT BETWEEN ('2171-1-14 22:23:8') AND ((-7::TINYINT)::TIMESTAMP));""");
+    }
+
+    @Test
+    public void issue3074() {
+        this.compileRustTestCase("""
+                CREATE MATERIALIZED VIEW v73_optimized AS SELECT 'a'>'헊';""");
+    }
+
+    @Test
+    public void issue3076() {
+        this.compileRustTestCase("""
+                CREATE MATERIALIZED VIEW v72_optimized AS SELECT SUM(NULL);""");
+    }
+
+    @Test
     public void issue3063() {
         this.getCCS("""
                 CREATE TABLE array_tbl(c1 INT ARRAY, c2 INT ARRAY);

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java
@@ -23,9 +23,13 @@ public class RegressionTests extends SqlIoTest {
 
     @Test
     public void issue3071() {
-        this.compileRustTestCase("""
+        this.statementsFailingInCompilation("""
                 CREATE TABLE t0(c0 char) with ('materialized' = 'true');
-                CREATE MATERIALIZED VIEW v100_optimized AS (SELECT * FROM t0 WHERE (t0.c0) NOT BETWEEN ('Y') AND ('䤈'));""");
+                CREATE MATERIALIZED VIEW v100_optimized AS (SELECT * FROM t0 WHERE (t0.c0) NOT BETWEEN ('Y') AND ('䤈'));""",
+                "Failed to encode");
+        this.statementsFailingInCompilation("""
+                CREATE MATERIALIZED VIEW v73_optimized AS SELECT 'a'>'헊';""",
+                "Failed to encode");
     }
 
     @Test
@@ -44,12 +48,6 @@ public class RegressionTests extends SqlIoTest {
         this.compileRustTestCase("""
                 CREATE MATERIALIZED VIEW v10_optimized AS
                 SELECT SOME(('2059-10-5 5:3:37') NOT BETWEEN ('2171-1-14 22:23:8') AND ((-7::TINYINT)::TIMESTAMP));""");
-    }
-
-    @Test
-    public void issue3074() {
-        this.compileRustTestCase("""
-                CREATE MATERIALIZED VIEW v73_optimized AS SELECT 'a'>'헊';""");
     }
 
     @Test

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/VariantTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/VariantTests.java
@@ -47,9 +47,7 @@ public class VariantTests extends BaseSQLTests {
     public DBSPCompiler testCompiler() {
         // Do not optimize, esp in Calcite
         CompilerOptions options = this.testOptions(false, false);
-        DBSPCompiler compiler = new DBSPCompiler(options);
-        this.prepareInputs(compiler);
-        return compiler;
+        return new DBSPCompiler(options);
     }
 
     @Override

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/tools/BaseSQLTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/tools/BaseSQLTests.java
@@ -63,6 +63,7 @@ public class BaseSQLTests {
 
     public CompilerCircuitStream getCCS(String sql) {
         DBSPCompiler compiler = this.testCompiler();
+        this.prepareInputs(compiler);
         compiler.compileStatements(sql);
         return new CompilerCircuitStream(compiler);
     }

--- a/sql-to-dbsp-compiler/build.sh
+++ b/sql-to-dbsp-compiler/build.sh
@@ -19,7 +19,7 @@ else
     # hasn't been merged yet
     CALCITE_REPO="https://github.com/mihaibudiu/calcite.git"
     CALCITE_BRANCH="variant"
-    CALCITE_NEXT_COMMIT="5fc918661c44e8a9663a4456322316cf26b27cd9"
+    CALCITE_NEXT_COMMIT="6317e1cbf1914390a6c1062c8563187b65468729"
 fi
 CALCITE_NEXT="1.39.0"
 CALCITE_CURRENT="1.38.0"

